### ISSUE TITLE
[STORM-1549] Add support for extending tuple tree timeout

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#5384]
<p>During the discussion of <a href="https://github.com/apache/storm/pull/700" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/700</a> the issue of allowing timeout extension in case of unavailable external components (such as a web service) came up.</p>

<p>The current implementation makes tuples fail at a set interval, regardless of whether or not replaying them is necessary. This can be irritating in topologies that emit to multiple services, since one hanging service will cause replays to hit all the working services as well.</p>

<p>I suggest adding a resetTimeout function to IOutputCollector, which will make the relevant ackers and spouts reinsert the tuple tree information in their pending maps. </p>

<p>The intended usage is that a bolt can call this function on an interval if it needs to delay expiration, for example if it needs to retry calling a web server a few times. It may also be useful for slow topologies that want Storm to detect hanging/dropped tuples faster than the max expected complete latency of the topology.</p>

---
<details><summary><i>Originally reported by <a title="Stig Rohde Døssing" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=srdo">srdo</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-1549" target="_blank">Add support for extending tuple tree timeout</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Stig Rohde Døssing" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=srdo">srdo</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Minor
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2016-03-16T14:39:26+00:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user srdo opened a pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1107" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1107</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-1549" title="Add support for extending tuple tree timeout" class="issue-link" data-issue-key="STORM-1549"><del>STORM-1549</del></a>: Add support for resetting tuple timeout from bolts via the OutputCollector</p>

<p>    See <a href="https://issues.apache.org/jira/browse/STORM-1549" class="external-link" rel="nofollow">https://issues.apache.org/jira/browse/STORM-1549</a>.</p>

<p>    Other than implementing tuple timeout reset, I made a minor change to assert-loop in integration_test.clj so it doesn't hang on failing tests. It should now fail the test after 10 seconds. I hope that's okay.</p>

<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/srdo/storm" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/srdo/storm</a> 1.x-branch</p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/storm/pull/1107.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1107.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #1107</p>

<hr />
<p>commit 406052cdc7138046a79104fc5c6f72212415f7f2<br/>
Author: Stig Rohde Døssing <generalbas.srd@gmail.com><br/>
Date:   2016-02-14T01:39:42Z</p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-1549" title="Add support for extending tuple tree timeout" class="issue-link" data-issue-key="STORM-1549"><del>STORM-1549</del></a>: Add support for resetting tuple timeout from bolts via the OutputCollector</p>

<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user srdo commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1107#issuecomment-190826744" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1107#issuecomment-190826744</a></p>

<p>    Anyone have any feedback on this?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1107#issuecomment-190858491" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1107#issuecomment-190858491</a></p>

<p>    @srdo Sorry I have been really behind on reviews.  The code and the concept seem sound.  I have not thought through all of the implications of this, but as of now I am +1 for this.</p>

<p>    The only real issue is that we need a pull request/feature to go into master before we can do something for 1.x-branch.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user srdo commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1107#issuecomment-190881185" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1107#issuecomment-190881185</a></p>

<p>    Great. I'll make another PR when I've made these changes on master. Should I close this one?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1107#issuecomment-190905862" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1107#issuecomment-190905862</a></p>

<p>    @srdo If you want to get the changes into 1.x keep this around.  master and 1.x have diverged enough that it is not a simple sherry-pick any more for this part of the code.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user srdo opened a pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1174" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1174</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-1549" title="Add support for extending tuple tree timeout" class="issue-link" data-issue-key="STORM-1549"><del>STORM-1549</del></a>: <span class="error">&#91;master&#93;</span> Add support for resetting tuple timeout from bolts via the OutputCollector</p>

<p>    This is a port of <a href="https://github.com/apache/storm/pull/1107" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1107</a> to master.</p>

<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/srdo/storm" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/srdo/storm</a> master</p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/storm/pull/1174.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1174.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #1174</p>

<hr />
<p>commit d36be51a39abb03ac47e01eb2e1fda31f9f9110b<br/>
Author: Stig Rohde Døssing <generalbas.srd@gmail.com><br/>
Date:   2016-02-14T01:39:42Z</p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-1549" title="Add support for extending tuple tree timeout" class="issue-link" data-issue-key="STORM-1549"><del>STORM-1549</del></a>: Add support for resetting tuple timeout from bolts via the OutputCollector</p>

<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user srdo commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1107#issuecomment-190949609" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1107#issuecomment-190949609</a></p>

<p>    @revans2 Sure, I'll leave this open then. Here's the master version of this PR <a href="https://github.com/apache/storm/pull/1174" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1174</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user abhishekagarwal87 commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1174#discussion_r54685047" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1174#discussion_r54685047</a></p>

<p>    &#8212; Diff: storm-core/src/jvm/org/apache/storm/daemon/Acker.java &#8212;<br/>
    @@ -100,6 +101,8 @@ public void execute(Tuple input) {<br/>
 }<br/>
 curr.failed = true;<br/>
 pending.put(id, curr);<br/>
    +} else if(ACKER_RESET_TIMEOUT_STREAM_ID.equals(streamId)) {<br/>
    &#8212; End diff &#8211;</p>

<p>    minor space around if</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user abhishekagarwal87 commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1174#discussion_r54685209" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1174#discussion_r54685209</a></p>

<p>    &#8212; Diff: storm-core/src/jvm/org/apache/storm/daemon/Acker.java &#8212;<br/>
    @@ -100,6 +101,8 @@ public void execute(Tuple input) {<br/>
 }<br/>
 curr.failed = true;<br/>
 pending.put(id, curr);<br/>
    +} else if(ACKER_RESET_TIMEOUT_STREAM_ID.equals(streamId)) {<br/>
    &#8212; End diff &#8211;</p>

<p>    This introduces a NPE scenario because earlier in all the if blocks, if curr was null, it was getting initialized. Now curr.spoutTask will throw NPE if reset timeout comes before init or after init times out. This can be avoided by adding below just before `pending.put(id, curr)`<br/>
    ```<br/>
    if (curr == null) </p>
{
 curr = new AckObject();
    }
<p>    ```</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user abhishekagarwal87 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1174#issuecomment-191094570" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1174#issuecomment-191094570</a></p>

<p>    Looks good to me after the possible NPE is fixed. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1174#discussion_r54730661" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1174#discussion_r54730661</a></p>

<p>    &#8212; Diff: storm-core/src/jvm/org/apache/storm/task/OutputCollector.java &#8212;<br/>
    @@ -218,6 +218,16 @@ public void fail(Tuple input) </p>
{
     _delegate.fail(input);
 }

<p>    +    /**<br/>
    +    * Resets the message timeout for any tuple trees to which the given tuple belongs.<br/>
    +    * The timeout is reset to Config.TOPOLOGY_MESSAGE_TIMEOUT_SECS.<br/>
    +    * @param input the tuple to reset timeout for<br/>
    +    */<br/>
    &#8212; End diff &#8211;</p>

<p>    Can we add to the javadoc that this should be used sparingly?  This is not a cheap operation, so having someone do it for each tuple they receive is very expensive.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1174#discussion_r54730716" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1174#discussion_r54730716</a></p>

<p>    &#8212; Diff: storm-core/src/jvm/org/apache/storm/topology/BasicOutputCollector.java &#8212;<br/>
    @@ -52,6 +52,10 @@ public void emitDirect(int taskId, List<Object> tuple) </p>
{
     emitDirect(taskId, Utils.DEFAULT_STREAM_ID, tuple);
 }

<p>    +    public void resetTimeout(Tuple tuple){<br/>
    &#8212; End diff &#8211;</p>

<p>    Can we put the javadoc here too for resetTimeout from outputCollector?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1174#issuecomment-191264712" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1174#issuecomment-191264712</a></p>

<p>    Overall things look good.  Just a few minor nits about javadocs and what @abhishekagarwal87 pointed out.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1174#discussion_r54731300" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1174#discussion_r54731300</a></p>

<p>    &#8212; Diff: storm-core/src/jvm/org/apache/storm/daemon/Acker.java &#8212;<br/>
    @@ -100,6 +101,8 @@ public void execute(Tuple input) {<br/>
 }<br/>
 curr.failed = true;<br/>
 pending.put(id, curr);<br/>
    +} else if(ACKER_RESET_TIMEOUT_STREAM_ID.equals(streamId)) {<br/>
    &#8212; End diff &#8211;</p>

<p>    Yes this is very true, in the clj code the entire processing was wrapped in a null check for curr.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user srdo commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1174#issuecomment-191326068" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1174#issuecomment-191326068</a></p>

<p>    The NPE should be fixed, I've added the missing space to the if, and added a comment to the javadoc about reset timeout being an expensive operation. The comment has also been added to the 1.x-branch version</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user rsltrifork commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1174#issuecomment-194200201" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1174#issuecomment-194200201</a></p>

<p>    As all review comments are addressed, will it be merged soon? <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/><br/>
    We need this feature in the IoT storm topology we're building, where bolts send to unstable external systems and we need at-least-once consistency, so dropping tuples on tuple timeout is bad for us.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user srdo commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1174#issuecomment-197008299" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1174#issuecomment-197008299</a></p>

<p>    @revans2 @abhishekagarwal87 Sorry, didn't notice this had merge conflicts. It should be good to go now though <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user srdo commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1107#issuecomment-197016989" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1107#issuecomment-197016989</a></p>

<p>    The test failure looks unrelated to this PR, was also happening on 1.x-branch <a href="https://travis-ci.org/apache/storm/jobs/115948898" class="external-link" target="_blank" rel="nofollow noopener">https://travis-ci.org/apache/storm/jobs/115948898</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user abhishekagarwal87 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1174#issuecomment-197188509" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1174#issuecomment-197188509</a></p>

<p>    Looks good to me. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user asfgit closed the pull request at:</p>

<p>    <a href="https://github.com/apache/storm/pull/1107" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1107</a></p>

**Comment by jira-importer** [comment]
<i>[Duplicates: <a href="https://github.com/apache/storm/issues?q=in%3Atitle%20STORM-171">STORM-171</a>]</i>
